src/emqx_channel.erl

```diff
@@ -1145,18 +1145,26 @@ interval(will_timer, #channel{will_msg = WillMsg}) ->
 %%--------------------------------------------------------------------
 
 -spec(terminate(any(), channel()) -> ok).
-terminate(_, #channel{conn_state = idle}) -> ok;
+terminate(_Reason, #channel{conn_state = idle} = _Channel) ->
+    ?tp(channel_terminated, #{channel => _Channel, reason => _Reason}),
+    ok;
 terminate(normal, Channel) ->
     run_terminate_hook(normal, Channel);
 terminate({shutdown, Reason}, Channel)
-  when Reason =:= kicked; Reason =:= discarded; Reason =:= takeovered ->
+  when Reason =:= kicked;
+       Reason =:= discarded;
+       Reason =:= takeovered;
+       Reason =:= not_authorized ->
```

Comment:
The judgment here may not be enough? For example, when we set `acl_deny_action = disconnect` in emqx.conf, the connection will also be shutdown with a `not_authorized` error since publish/subscribe rejected. But the Will-Message should be sent in this case.

Reply:
Thanks for pointing that out!

I've changed the logic to consider the reason, the existence of a will message and the connection state to decide if the message should be published.